### PR TITLE
Fix GPU OOM issues

### PR DESCRIPTION
**Pull request by jmiller656**
Hi 👋 

You guys have a pretty cool project here. I ran your notebook and saw you guys mentioned a lot of OOM issues. It looks like the big problem wasn't RAM, but instead was loading the entire dataset onto GPU. I made a couple of small tweaks to the training notebook that should alleviate this. I was able to run the "extended training mode" of this on GPU on my laptop (~4GB GPU memory, 8GB RAM, MS Surface Book 2). So this should work on colab as well!

### CODE REVIEW

**Comment by mishushakov** [comment]
hey, thanks for the PR
i've tried running the notebook on Google Colab, but it restarts because of OOM

Screenshot 2021-01-31 at 15 01 21

any ideas? thanks

**Comment by mishushakov** [comment]
diff for @GuitarML

```diff
diff --git a/index.py b/index.py
index 4c74814..576447c 100644
--- a/index.py
+++ b/index.py
@@ -30,10 +30,9 @@ import h5py
 # EDIT THIS SECTION FOR USER INPUTS
 #
 name = 'test'
-in_file = 'ts9_test1_in_FP32.wav'
-out_file = 'ts9_test1_out_FP32.wav'
+in_file = 'data/ts9_test1_in_FP32.wav'
+out_file = 'data/ts9_test1_out_FP32.wav'
 epochs = 1
-split_data=4 # **Increase this to reduce RAM usage **
 
 train_mode = 0     # 0 = speed training, 
                    # 1 = accuracy training 
@@ -129,7 +128,7 @@ model.add(Conv1D(conv1d_filters, 12,strides=conv1d_strides, activation=None, pad
 model.add(LSTM(hidden_units))
 model.add(Dense(1, activation=None))
 model.compile(optimizer=Adam(learning_rate=learning_rate), loss=error_to_signal, metrics=[error_to_signal])
-print(model.summary())
+model.summary()
 
 # Load and Preprocess Data ###########################################
 in_rate, in_data = wavfile.read(in_file)
@@ -138,52 +137,15 @@ out_rate, out_data = wavfile.read(out_file)
 X_all = in_data.astype(np.float32).flatten()  
 X_all = normalize(X_all).reshape(len(X_all),1)   
 y_all = out_data.astype(np.float32).flatten() 
-y_all = normalize(y_all).reshape(len(y_all),1)   
+y_all = normalize(y_all).reshape(len(y_all),1)
 
-# If splitting the data for training, do this part
-if split_data > 1:
-    num_split = len(X_all) // split_data
-    X = X_all[0:num_split*split_data]
-    y = y_all[0:num_split*split_data]
-    X_data = np.split(X, split_data)
-    y_data = np.split(y, split_data)
+y_ordered = y_all[input_size-1:] 
+indices = np.arange(input_size) + np.arange(len(X_all)-input_size+1)[:,np.newaxis] 
+x_ordered = np.take(X_all, indices)[:,:, np.newaxis]
 
-    # Perform training on each split dataset
-    for i in range(len(X_data)):
-        print("\nTraining on split data " + str(i+1) + "/" +str(len(X_data)))
-        X_split = X_data[i]
-        y_split = y_data[i]
-
-        y_ordered = y_split[input_size-1:] 
-
-        indices = np.arange(input_size) + np.arange(len(X_split)-input_size+1)[:,np.newaxis] 
-        X_ordered = tf.gather(X_split,indices) 
-
-        shuffled_indices = np.random.permutation(len(X_ordered)) 
-        X_random = tf.gather(X_ordered,shuffled_indices)
-        y_random = tf.gather(y_ordered, shuffled_indices)
-
-        # Train Model ###################################################
-        model.fit(X_random,y_random, epochs=epochs, batch_size=batch_size, validation_split=0.2)  
-
-
-    model.save('models/'+name+'/'+name+'.h5')
-
-# If training on the full set of input data in one run, do this part
-else:
-    y_ordered = y_all[input_size-1:] 
-
-    indices = np.arange(input_size) + np.arange(len(X_all)-input_size+1)[:,np.newaxis] 
-    X_ordered = tf.gather(X_all,indices) 
-
-    shuffled_indices = np.random.permutation(len(X_ordered)) 
-    X_random = tf.gather(X_ordered,shuffled_indices)
-    y_random = tf.gather(y_ordered, shuffled_indices)
-
-    # Train Model ###################################################
-    model.fit(X_random,y_random, epochs=epochs, batch_size=batch_size, validation_split=test_size)    
-
-    model.save('models/'+name+'/'+name+'.h5')
+# Train Model ###################################################
+model.fit(x_ordered,y_ordered, epochs=epochs, batch_size=batch_size, validation_split=test_size, shuffle=True)    
+model.save('models/'+name+'/'+name+'.h5')
 
 # Run Prediction #################################################
 print("Running prediction..")
@@ -193,7 +155,7 @@ y_the_rest, y_last_part = np.split(y_all, [int(len(y_all)*.8)])
 x_the_rest, x_last_part = np.split(X_all, [int(len(X_all)*.8)])
 y_test = y_last_part[input_size-1:] 
 indices = np.arange(input_size) + np.arange(len(x_last_part)-input_size+1)[:,np.newaxis] 
-X_test = tf.gather(x_last_part,indices) 
+X_test = np.take(x_last_part,indices)[:, :, np.newaxis]
 
 prediction = model.predict(X_test, batch_size=batch_size)
 
@@ -208,3 +170,6 @@ grp = f.create_group("info")
 dset = grp.create_dataset("input_size", (1,), dtype='int16')
 dset[0] = input_size
 f.close()
+
+import IPython.display as ipd
+ipd.Audio('models/'+name+'/y_pred.wav')

```

**Comment by jmiller656** [comment]
😅 oops yeah, looks like that still doesn't quite work on colab, does it?

🤔I think I'll try and take a streaming approach then and update this PR when I get it running on colab

**Comment by mishushakov** [comment]
cool, np!

GuitarLSTM already can "survive" input size 50-80 on colab
with `--split_data ` even more, my best was 250

any improvements are very welcome, see #6
thanks

**Comment by jmiller656** [comment]
Got it running in colab with `input_size=1500` right now with no OOM. Going to clean some stuff up and push tomorrow night. 

TL;DR: storing the indices arrays takes up a ton of RAM. I fix this by using the [keras Sequence class](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence)

**Comment by mishushakov** [comment]
awesome, looking forward to try it out!

**Comment by jmiller656** [comment]
Should be good to go now, feel free to check it out!

One thing worth mentioning though, if you use the `error_to_signal` loss for this model, you'll probably notice the loss becomes very large. This is because of the zeroes in the data. I've made an issue for that here: #10 

My thoughts on this are that you can probably fairly safely replace the `error_to_signal` loss with `mse` instead. They are pretty similar aside from the denominator and the `pre_emphasis_filter`. Empirically, I've found `mse` seems to work and sound just as good.

**Comment by GuitarML** [comment]
@jmiller656 The use of the Sequence class seems to fix the RAM problem, but I don't think it's actually training the model correctly for me.  When you listened to the pred.wav file did you get a good sound?  You are correct about using MSE, you get a good result using that as well.  There's actually a more recent paper out of Aalto University that uses different methods on the loss function, so I might be changing that part anyway.

**Comment by jmiller656** [comment]
Hmm, just looked into this. I found the same problem. Looks like that's only the case if you use `error_to_signal` as the loss. The instability of the loss is probably making the gradients explode and hurting the network more than it helps. Switching to `mse` still seems to work well though, and if we wanted a band-aid solution for the `error_to_signal` loss, we could probably clip the value before returning it.

Could you send me that paper? I'm curious about what they do, happy to try it out and let you know how it goes

**Comment by GuitarML** [comment]
Actually now that I look at the date it might not be new, but it’s one I hadn’t seen before and it goes into detail on the loss function.

https://arxiv.org/pdf/1911.08922v1.pdf

Can you update the colab code with the working MSE when you get a chance? Id like to run some comparison tests between that and the original. Good work on writing the Sequence data loader class! It’s much cleaner than splitting the data into big chunks.

**Comment by jmiller656** [comment]
Yup! This should have `mse` loss now

**Comment by GuitarML** [comment]
@jmiller656 Initial listening tests sound very promising! To my ears it models the ts9 about the same as the original loss function. When using something with heavier distortion it sounds pretty good, but picks up on different qualities of the sound.  I'm going to do more testing but I'm also going to merge the pull request using MSE, because for Colab it's simply a much better solution for loading the audio data. Well done!  I'm learning that the choice of loss function is one of the most important factors, so I'm going to do some more research and probably make some changes across all the repos based on what I find.  You seem to have a good understanding of this area, so any further help is greatly appreciated.

**Comment by GuitarML** [comment]
@jmiller656 I ran some tests, and the original error to signal calculation produces much better sounding models for highly distorted signals. I created this issue to investigate using your dataloader class with the original loss function. 

https://github.com/GuitarML/GuitarLSTM/issues/11

Any thoughts on how to make this work, and why the original loss function is so unstable with the dataloader class? It seems to work fine in most cases when running on the entire or split dataset, except for the 'inf' issue.

**Comment by mishushakov** [comment]
i have to admit, when training on original dataset the resulting model produces too much treble, sounds too bright on the high end

i'm not sure if this is related to the changes in the commit or just a issue in general
thanks

**Comment by jmiller656** [comment]
Just responded with some thoughts in #11,
